tests/test_cli.py

```diff
@@ -36,224 +32,330 @@ def limited_env(monkeypatch):
             monkeypatch.setenv(variable, value)
 
 
-@pytest.mark.usefixtures("runner")
-class TestConnect(object):
-    def test_help(self):
-        result = self.runner.invoke(
-            connect, ["--help"], standalone_mode=False, catch_exceptions=False
-        )
-        assert result.exit_code == 0
-
-    def test_no_arguments_exits_with_nonzero_code(self, clean_env):
-        result = self.runner.invoke(connect)
-        assert result.exit_code != 0
-
-    @patch("fonz.cli.Fonz", autospec=True)
-    def test_with_command_line_args_only(self, mock_client, clean_env):
-        result = self.runner.invoke(
-            connect,
-            [
-                "--base-url",
-                TEST_BASE_URL,
-                "--client-id",
-                "CLIENT_ID_CLI",
-                "--client-secret",
-                "CLIENT_SECRET_CLI",
-            ],
-            standalone_mode=False,
-            catch_exceptions=False,
-        )
-        mock_client.assert_called_once_with(
-            TEST_BASE_URL, "CLIENT_ID_CLI", "CLIENT_SECRET_CLI", 19999, "3.0"
+@pytest.fixture()
+def parser():
+    parser = create_parser()
+    return parser
+
+
+def test_help(parser,):
+    with pytest.raises(SystemExit) as cm:
+        with patch.object(sys, "argv", ["fonz", "--help"]):
+            main()
+            assert cm.value.code == 0
+
+
+@patch("fonz.cli.connect", autospec=True)
+def test_connect_with_base_cli(mock_connect, clean_env):
+    with patch.object(
+        sys,
+        "argv",
+        [
+            "fonz",
+            "connect",
+            "--base-url",
+            "cli_url",
+            "--client-id",
+            "cli_client_id",
+            "--client-secret",
+            "cli_client_secret",
+        ],
+    ):
+        main()
+        mock_connect.assert_called_once_with(
+            "cli_url", "cli_client_id", "cli_client_secret", 19999, 3.0
         )
-        mock_client.return_value.connect.assert_called_once()
-        assert result.exit_code == 0
 
-    @patch("fonz.cli.Fonz", autospec=True)
-    def test_with_env_vars_only(self, mock_client, env):
-        result = self.runner.invoke(
-            connect, standalone_mode=False, catch_exceptions=False
+
+@patch("fonz.cli.connect", autospec=True)
+def test_connect_with_full_cli(mock_connect, clean_env):
+    with patch.object(
+        sys,
+        "argv",
+        [
+            "fonz",
+            "connect",
+            "--base-url",
+            "cli_url",
+            "--client-id",
+            "cli_client_id",
+            "--client-secret",
+            "cli_client_secret",
+            "--port",
+            "272727",
+            "--api-version",
+            "3.1",
+        ],
+    ):
+        main()
+        mock_connect.assert_called_once_with(
+            "cli_url", "cli_client_id", "cli_client_secret", "272727", "3.1"
```

Comment:
Kinda. But have changed it so that we always get strings, which I think is what we want?

Reply:
I ended up going with float and int in my `feature/misc` branch since I needed to be able to sort them, but we can merge this as is and deal with that change later.